org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/IndexUtils.java

```diff
@@ -0,0 +1,144 @@
+
+                        boolean needDeleteLocalIndex = false;
+                        File localIndexFile = localIndexLocation.getIndexFile();
+                        if (localIndexLocation.exists()) {
+                            File sharedIndexFile = sharedIndexLocation.getIndexFile();
+                            if (sharedIndexFile == null || localIndexFile == null) {
+                                continue;
+                            }
+
+                            if (indexManager.getIndex(sharedIndexLocation) != null) {
+                                // current classpath entry is using the shared index, delete local index directly.
+                                needDeleteLocalIndex = true;
+                            } else {
+                                Index localIndex = indexManager.getIndex(localIndexLocation);
+                                if (localIndex == null) {
+                                    continue;
+                                }
+
+                                ReadWriteMonitor monitor = localIndex.monitor;
+                                if (monitor == null) { // index got deleted since acquired
+                                    continue;
+                                }
+    
+                                File tmpFile = new File(sharedIndexFile.getParent(), sharedIndexFile.getName() + "." + System.currentTimeMillis());
+                                try {
+                                    monitor.enterRead();
+                                    mkdirsFor(tmpFile);
+                                    Files.copy(localIndexFile.toPath(), tmpFile.toPath(), LinkOption.NOFOLLOW_LINKS, StandardCopyOption.REPLACE_EXISTING);
+                                    Files.deleteIfExists(sharedIndexFile.toPath());
+                                    needDeleteLocalIndex = tmpFile.renameTo(sharedIndexFile);
+                                } catch (IOException e) {
+                                    JavaLanguageServerPlugin.logError(String.format("Failed to copy the local index %s to the shared index %s: %s",
+                                        localIndexFile.getName(), sharedIndexFile.getName(), e.getMessage()));
+                                } finally {
+                                    monitor.exitRead();
+                                    try {
+                                        Files.deleteIfExists(tmpFile.toPath());
+                                    } catch (IOException e) {
+                                        // do nothing
+                                    }
+                                }
+                            }
+                        }
+
+                        if (needDeleteLocalIndex) {
+                            try {
+                                // Remove the local index after it has been copied to the shared index location.
+                                Files.deleteIfExists(localIndexFile.toPath());
```

Comment:
What will happen after we delete the location index file, but the location mappings are not updated in the index manager?

Reply:
Not completely sure, but I think it would only switch over to the shared index, if a change is detected on the index (doesn't really happen for a jar), or if we find a way to manually clear it out.

I see there's a `clearCache()` in the index manager. Maybe that's enough to re-read the entries without re-writing anything. I don't remember ever testing the case of copying over an index from the "local" area to the "shared" area after import has already begun (to see if the shared area takes over).

Reply:
> What will happen after we delete the location index file, but the location mappings are not updated in the index manager?

This is a really good question.

I've tried it and it can cause the cached indexes to be unusable. This is because the search engine will use the cached **Index/DiskIndex** instances in memory for results. Simply deleting the index file may break the cached DiskIndex instance on that file, because the [query operation on the DiskIndex](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/7eeb9b4342d4d83f3a5a1d078a4e4dff0ab83bb2/org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/index/DiskIndex.java#L169) is lazily loading the index data on disk and may get a FileNotFound error.

I came up with a conservative way to solve this problem. The first time we only copy the local index to the shared location, but without deleting the local copy. The next time when you reload the language server, it is safe to delete the local indexes because the language server will automatically switch to the shared index and local indexes become unused.

From what I've tried, the new approach works well.